src/language/lexer.rs

```diff
@@ -2,26 +2,26 @@
 use std::str::CharIndices;
 use std::iter::Peekable;
 
-#[derive(Debug, PartialEq, Eq)]
+#[derive(Debug, PartialEq)]
 pub enum Token<'a> {
-  Eof,
-  Bang,
-  Dollar,
-  LeftParen,
-  RightParen,
-  Spread,
-  Colon,
-  Equals,
-  At,
-  LeftBracket,
-  RightBracket,
-  LeftBrace,
-  RightBrace,
-  Pipe,
-  Name(&'a str),
-  Int(&'a str),
-  Float(&'a str),
-  String(&'a str),
+  Eof(usize, usize),
+  Bang(usize, usize),
+  Dollar(usize, usize),
+  LeftParen(usize, usize),
+  RightParen(usize, usize),
+  Spread(usize, usize),
+  Colon(usize, usize),
+  Equals(usize, usize),
+  At(usize, usize),
+  LeftBracket(usize, usize),
+  RightBracket(usize, usize),
+  LeftBrace(usize, usize),
+  RightBrace(usize, usize),
+  Pipe(usize, usize),
+  Name(usize, usize, &'a str),
+  Int(usize, usize, &'a str),
+  Float(usize, usize, &'a str),
```

Comment:
Agree to rename Int, Float, and String to IntValue, FloatValue, and StringValue.
for the Punctuator token, how about Punctuator(&str, usize, usize) ?
later it can pattern match like Punctuator('!', _, _)
I also prefer put the location information at the end.